### DISCUSSION TITLE
How to convert apng into webp?

### DISCUSSION THREAD

**Discussion by FurkanGozukara** [dlemstra/Magick.NET#962] [Help]
I have tried like this but apng animation is lost

Also I am looking for settings to lossless yet maximum compression. I don't care about CPU time speed etc. I need lossless and maximum compression. how to achieve that?

So I have a lot of PNG files and some APNG files

I will turn all of them into WEBP files

Yet I still want to keep APNG animation as well

Here example apng to test : https://static.pokemonpets.com/images/npcTrainers/trainer15125down.png

```
         using (var image = new MagickImage(@"path\Debug\a.png"))
            {
                image.Quality = 100;
                image.Settings.SetDefine(MagickFormat.WebP, "lossless", true);
                image.Settings.SetDefine(MagickFormat.WebP, "method", "6");
                image.Write("a.webp");
            }
```

**Comment by dlemstra** [comment]
If you want to read an image that contains multiple frames you will need to use the `MagickImageCollection` instead:

```C#
using (var images = new MagickImageCollection(@"path\Debug\a.png"))
{
    var defines = new WebPWriteDefines
    {
        Lossless = true,
        Method = 6,
    };

    images.Write("a.webp", defines);
}
```

And if you want to read an animated PNG file you will need to have `ffmpeg.exe` in your `%PATH%`. This is required to read an animated PNG fil.e

**Reply by FurkanGozukara** [reply]
Thank you very much for the answer

I have made it generate animated image but the animated image has black background and its size is like 10x so something is clearly wrong

This is the image i am testing : https://static.pokemonpets.com/images/npcTrainers/trainer15202down.png

It only reads animated PNG if you only provide format

So here my code

![image](https://user-images.githubusercontent.com/19240467/122644712-57abcb80-d11f-11eb-8f76-2755759dd4e6.png)


```
      FileInfo img_apng = new FileInfo(srWorkingDirectory + @"\test\trainer15202down.png");
            using (var images = new MagickImageCollection(img_apng, MagickFormat.APng))
            {
                var defines = new WebPWriteDefines
                {
                    Lossless = true,
                    Method = 6
                };

                images.Write(srWorkingDirectory + @"\test\trainer15202down.webp", defines);
            }
```

And here the generated output which is very wrong

![image](https://user-images.githubusercontent.com/19240467/122644690-377c0c80-d11f-11eb-9623-8930c5ac51a2.png)


[trainer15202down.zip](https://github.com/dlemstra/Magick.NET/files/6680801/trainer15202down.zip)

**Reply by dlemstra** [reply]
You probably need to rename the file to `.apng` to let ImageMagick know it is an animated png file.

*EDIT*: It should be `apng:filename.png` or as in the example below add the format as an argument to the constructor.

**Reply by FurkanGozukara** [reply]
@dlemstra apng extension doesn't work I just tested. I had to provide MagickFormat.APng 

Now it reads however reading it totally incorrectly

It shows that image have 607 frames 

![image](https://user-images.githubusercontent.com/19240467/122644875-14059180-d120-11eb-8712-9c2d12786c1a.png)

However in reality the image has only 66 frames

![image](https://user-images.githubusercontent.com/19240467/122644889-27b0f800-d120-11eb-848d-58507cb2a599.png)


Therefore the output file is too big and have black background

Any ideas?

Thank you very much

**Reply by FurkanGozukara** [reply]
@dlemstra Ok I think it might be related to ImageMagick  https://github.com/ImageMagick/ImageMagick/issues/3847

**Reply by dlemstra** [reply]
A new release that includes the class `VideoReadDefines` has just been published with a `VideoSync` property that can be used to reduce the number of exported frames.

**Reply by FurkanGozukara** [reply]
@dlemstra thank you very much. can you write an example?

**Reply by dlemstra** [reply]
Sure, you can do this:

```C#
var readDefines = new VideoReadDefines(MagickFormat.APng)
{
    VideoSync = VideoSync.Vfr
};

FileInfo img_apng = new FileInfo(srWorkingDirectory + @"\test\trainer15202down.png");
using (var images = new MagickImageCollection(img_apng, readDefines))
{
}
```

**Reply by FurkanGozukara** [reply]
@dlemstra awesome ty very much. 1 more question, how can I get complexity of an image? so that I can apply different filtering. I am sure there is no universal complexity but If I can get accurate relative complexity of images, I can evaluate and set filters. Can you show me how to get an image complexity?

**Reply by FurkanGozukara** [reply]
@dlemstra a command like this how to execute with magick net? magick   dragon.png   -alpha off   -statistic StandardDeviation 5x5   -fill White +opaque Black   -format "%[fx:mean]\n

**Reply by dlemstra** [reply]
Can you please open a new question for that instead? Your question is not related to this discussion.
